### PR TITLE
Adds generation of new PLE/PLP via ExecutionEngine

### LINKED ISSUE #805: Generation of new PLE and PLP via Execution Engine
- Return functionality that existed with old PLE / PLP tabs
- Make execution screen to be consistent with CC / Pathways

### PR DESCRIPTION
**Pull request by wivern**
closes #805

### CODE REVIEW

File: src/main/java/org/ohdsi/webapi/common/generation/AnalysisExecutionSupport.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+public abstract class AnalysisExecutionSupport extends AbstractDaoService {
+
+
+  protected AnalysisFile prepareAnalysisExecution(String packageName, String packageFilename, Integer analysisId) {
+      AnalysisFile execFile = new AnalysisFile();
+      execFile.setFileName("runAnalysis.R");
+      String[] paramNames = {"packageFile", "packageName", "analysisDir"};
+      String[] paramValues = {packageFilename, packageName, String.format("analysis_%d", analysisId)};
+      String script = SqlRender.renderSql(getExecutionScript(), paramNames, paramValues);
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/common/generation/AnalysisExecutionSupport.java]: I would add a comment here describing what's going on. When you look at it the first time, a question arises - what the heck SQL is doing here?

File: src/main/java/org/ohdsi/webapi/common/generation/GenerationUtils.java [hunk 2]
```diff
@@ -122,4 +141,44 @@ public SimpleJobBuilder buildJobForCohortBasedAnalysisTasklet(
...
+        Step runExecutionStep = stepBuilderFactory.get(analysisTypeName + ".startExecutionEngine")
+                .tasklet(runExecutionEngineTasklet)
+                .build();
+
+        Step waitCallbackStep = stepBuilderFactory.get(analysisTypeName + ".waitForCallback")
+                .tasklet(callbackTasklet)
+                .build();
+
+        JdbcTemplate jdbcTemplate = getSourceJdbcTemplate(source);
+        DropCohortTableListener dropCohortTableListener = new DropCohortTableListener(jdbcTemplate, transactionTemplate, sourceService, sourceAwareSqlRender);
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/common/generation/GenerationUtils.java]: Why would `dropCohortTableListener` be required here?

File: src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java [hunk 3]
```diff
@@ -308,41 +254,63 @@ public EstimationAnalysisImpl exportAnalysis(@PathParam("id") int id) {
...
+        try(ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            EstimationAnalysisImpl analysis = exportAnalysis(estimation);
+            analysis.setPackageName(packageName);
+            hydrateAnalysis(analysis, out);
+            analysisFile.setContents(out.toByteArray());
+        }
+        analysisFiles.add(analysisFile);
+        analysisFiles.add(prepareAnalysisExecution(packageName, packageFilename, analysisId));
+
+        JobParametersBuilder builder = prepareJobParametersBuilder(source, analysisId, packageName, packageFilename)
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java]: Do we really need Spring Batch here? I mean, the asynchronous execution is achieved by using Execution Engine, so why cannot we send the package directly (not via a tasklet) and avoid extra database views, extra tasklets and extra entities (so quite a lot of code could be removed)?

File: src/main/java/org/ohdsi/webapi/executionengine/controller/ScriptExecutionController.java [hunk 4]
```diff
@@ -92,7 +68,7 @@ public ScriptExecutionController(final ScriptExecutionService scriptExecutionSer
         this.sourceRepository = sourceRepository;
     }
 
-    @Path("execution/run")
+ /*   @Path("execution/run")
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/executionengine/controller/ScriptExecutionController.java]: Why did you leave the code commented out? why not remove it?

File: src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java (hunk 3 above)
Comment (chrisknoll) [reply src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java]: Depends: do you want this asycn action to be tracked by spring batch jobs?  Everything else is.

Comment (pavgra) [reply src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java]: Everywhere else we use Spring Batch to do async processing. Not just for things to be tracked. And here I do not see a reason for the Spring Batch usage

Comment (wivern) [reply src/main/java/org/ohdsi/webapi/estimation/EstimationServiceImpl.java]: But without Spring Batch it's unable to get PLP/PLE jobs on the Jobs page and receive notifications,

File: src/main/resources/db/migration/postgresql/V2.7.0.20190208164736__analysis_execution-add-job_id.sql [hunk 5]
```diff
@@ -0,0 +1,10 @@
+ALTER TABLE ${ohdsiSchema}.analysis_execution ADD IF NOT EXISTS job_execution_id BIGINT;
+ALTER TABLE ${ohdsiSchema}.analysis_execution DROP COLUMN IF EXISTS analysis_type;
```
Comment (pavgra) [inline src/main/resources/db/migration/postgresql/V2.7.0.20190208164736__analysis_execution-add-job_id.sql]: Why do you need `source_id` in `analysis_execution` if the source id of execution is saved in `batch_job_execution_params`? The same question for `analysis_execution.analysis_id` vs `CAST(plp_id_param.string_val AS INTEGER) prediction_id`, `analysis_execution.duration` and `analysis_execution.executed` vs `job.create_time` and `job.end_time`,  `analysis_execution.sec_user_id` and  `gen_info.created_by_id`, `analysis_execution.executionstatus` vs `job.status`. All of the `analysis_execution` fields seem to be redundant. So, all in all, why cannot we just drop the `analysis_execution`?

Comment (wivern) [reply src/main/resources/db/migration/postgresql/V2.7.0.20190208164736__analysis_execution-add-job_id.sql]: analysis_execution is needed for tracking EE analysis has completed

File: src/main/resources/db/migration/sqlserver/V2.7.0.20190225165752__estimation_gen_view_fixes.sql [hunk 6]
```diff
@@ -0,0 +1,22 @@
+ALTER VIEW ${ohdsiSchema}.estimation_analysis_generation as
```
Comment (pavgra) [inline src/main/resources/db/migration/sqlserver/V2.7.0.20190225165752__estimation_gen_view_fixes.sql]: Why cannot we leave only one, final migration? So why not to remove `V2.7.0.20190211120020__estimation_generation_view.sql`?

File: src/main/resources/db/migration/sqlserver/V2.7.0.20190225162439__rename_analysis_execution.sql [hunk 7]
```diff
@@ -0,0 +1,9 @@
+alter table ${ohdsiSchema}.analysis_execution drop constraint fk_ae_sec_user;
```
Comment (pavgra) [inline src/main/resources/db/migration/sqlserver/V2.7.0.20190225162439__rename_analysis_execution.sql]: 1. Why cannot we collapse it with `V2.7.0.20190208164736__analysis_execution-add-job_id.sql`?
2. Why `source` is still here? - duplicated in Spring batch jobs

File: src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java [hunk 8]
```diff
@@ -252,56 +201,45 @@ public String getExecutionStatus(Long executionId) {
...
-            temp = temp.replace("password = \"supersecret\"", "password = \""
-                    + dataSourceData.getPassword() + "\", "
-                    + "pathToDriver=\"/impala/\"");
+    @PostConstruct
+    public List<ExecutionEngineAnalysisStatus> findOutdatedAnalyses() {
 
-        } else {
-            temp = temp.replace("password = \"supersecret\"", "password = \"" + dataSourceData.getPassword() + "\"");
+        logger.info("Invalidating execution engine based analyses");
+        return analysisExecutionRepository.findByExecutionStatusIn(INVALIDATE_STATUSES);
```
Comment (pavgra) [inline src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java]: Where is the invalidation itself?

File: src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java [hunk 9]
```diff
@@ -211,10 +211,11 @@ public void updateAnalysisStatus(ExecutionEngineAnalysisStatus analysisExecution
     }
 
     @PostConstruct
-    public List<ExecutionEngineAnalysisStatus> findOutdatedAnalyses() {
+    public void findOutdatedAnalyses() {
```
Comment (chrisknoll) [inline src/main/java/org/ohdsi/webapi/executionengine/service/ScriptExecutionServiceImpl.java]: Strange name: if it's supposed to find something, why isn't it returning something?

**Comment by pavgra** [comment]
Remaining things:
- Redundant stacktrace when tasklet is FAILED
- Security (source) checks on results loading (list + archive)

**Pull request merged**